atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -479,43 +486,81 @@ public CheckAndSetCompatibility getCheckAndSetCompatibility() {
     @Override
     public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException {
         TableReference tableRef = request.table();
-        Table table = getTableMap(tableRef);
-        Cell cell = request.cell();
-        Optional<byte[]> oldValue = request.oldValue();
-        byte[] contents = request.newValue();
-
-        Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
-        if (oldValue.isPresent()) {
-            byte[] storedValue = table.entries.get(key);
-            boolean succeeded = Arrays.equals(storedValue, oldValue.get())
-                    && table.entries.replace(key, storedValue, copyOf(contents));
-            if (!succeeded) {
-                byte[] actual = table.entries.get(key); // Re-fetch, something may have happened between get and replace
-                throwCheckAndSetException(cell, tableRef, oldValue.get(), actual);
-            }
-        } else {
-            byte[] oldContents = putIfAbsent(table, key, contents);
-            if (oldContents != null) {
-                throwCheckAndSetException(cell, tableRef, null, oldContents);
+        runWithLockForWrites(tableRef, table -> {
+            Cell cell = request.cell();
+            Optional<byte[]> oldValue = request.oldValue();
+            byte[] contents = request.newValue();
+
+            Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
+            if (oldValue.isPresent()) {
+                byte[] storedValue = table.get(key);
+                boolean succeeded =
+                        Arrays.equals(storedValue, oldValue.get()) && table.replace(key, storedValue, copyOf(contents));
+                if (!succeeded) {
+                    byte[] actual = table.get(key); // Re-fetch, something may have happened between get and replace
+                    throwCheckAndSetException(cell, tableRef, oldValue.get(), actual);
+                }
+            } else {
+                byte[] oldContents = putIfAbsent(table, key, contents);
+                if (oldContents != null) {
+                    throwCheckAndSetException(cell, tableRef, null, oldContents);
+                }
             }
-        }
+            return null;
+        });
     }
 
+    /**
+     * This operation ensures atomicity by locking down the table. This ensures that there are no conflicts with other
+     * writes, and that reads see a consistent view of the world. The exception are reads that return iterators, which
+     * may see a partially applied multiCAS.
+     */
     @Override
     public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
-        throw new UnsupportedOperationException("InMemoryKvs does not support multi-checkAndSet operation!");
+        TableReference tableRef = multiCheckAndSetRequest.tableRef();
+        Table tableEntry = getTableMap(tableRef);
+        Lock lock = tableEntry.lock.writeLock();
+        lock.lock();
+        try {
```

Comment:
let's just refactor out a method `runWithWriteLocks` (:)) for this one? Just makes it easier to read.

Reply:
sure